src/ops/base.rs

```diff
@@ -67,8 +249,46 @@ impl Ingredient for Base {
                 rs: Records,
                 _: &DomainNodes,
                 state: &StateMap)
-                -> Records {
-        rs.into_iter()
+                -> Option<Records> {
+        // Write incoming records to log before processing them if we are a durable node.
+        let records_to_return;
+        let mut copy_rs = rs.clone();
+        match self.durability {
+            BaseDurabilityLevel::Buffered => {
+                // Perform a synchronous flush if one of the following conditions are met:
+                //
+                // 1. Enough time has passed since the last time we flushed.
+                // 2. Our buffer of write records reaches capacity.
+                let num_buffered_writes = self.buffered_writes.as_ref().unwrap().len();
+                let has_reached_capacity = num_buffered_writes + rs.len() >= BUFFERED_WRITES_CAPACITY;
+                let elapsed = self.last_flushed_at.unwrap().elapsed();
+                let has_reached_time_limit = elapsed >= Duration::from_millis(BUFFERED_WRITES_FLUSH_INTERVAL_MS);
+
+                if has_reached_capacity || has_reached_time_limit {
+                    self.buffered_writes.as_mut().unwrap().append(copy_rs.as_mut());
+
+                    let copy_buffered_writes = self.buffered_writes.as_mut().unwrap().clone();
+                    self.flush();
+
+                    // This returns everything that was buffered, plus the newly inserted records.
+                    records_to_return = Some(copy_buffered_writes);
+                } else {
+                    // Otherwise, buffer the records and don't send them downstream.
+                    self.buffered_writes.as_mut().unwrap().append(copy_rs.as_mut());
+                    return Some(Records::default())
```

Comment:
Yeah, that makes a lot more sense than Some(empty Vec).  Otherwise, `::on_input` could just return a regular records Vec.

So I updated callers to explicitly handle `Base::on_input` returning None in the case of buffered writes that are not propagated downstream.

Oddly enough I now get a panic from `Project::resolve_col` via `it_fails_to_resolve_literal`, but this seems unrelated since there's no explicit handling of `Base::on_input` results there?  Why would the check for `self.emit.is_some() && col >= self.emit.as_ref().unwrap().len()` be true if `Base::on_input` can now return `None`?